### DISCUSSION TITLE
Umbraco.ColorPicker - Have a different color for each domain

### DISCUSSION THREAD

**Discussion by clementfavre** [umbraco/Umbraco-CMS#17216] [Features and ideas]
Hello everyone,
_Context_: I have a Umbraco site (13+) that uses widgets. These widgets have a background color property.
In Umbraco, I have several sub-sites that share the same widgets, but these sub-sites have different colors.

Current Picker:
![image](https://github.com/user-attachments/assets/69cf2c5b-fa7d-4956-ad9c-4b4c60b5fbef)

Choice of colors in the widget:
![image](https://github.com/user-attachments/assets/b6853acf-71e4-4e5a-bfd6-7a2fe5accbc1)

How can I distinguish the colors in relation to the sub-sites? 
It might be useful to link a color to a domain, so that the color picker is different for each sub-domain.

**Comment by nackler** [comment]
I did something similar with domain-specific settings -- not colors, but the idea is the same.

Add a settings document (no template) in the root of your site where color/subsite relationships are configured in the backoffice as pairs of color/domain-name values.  Then when your widget is rendering, call a utility from the view that finds that settings document in the content tree, then loops through the configured colors trying to match the domain from the http request with the domain-name in the settings document.


If, for some reason you need to look up the assigned domain bindings of the subsite nodes, you can do that using an injected umbraco IDomainService.  

Let me know if you need examples of any of this.

**Comment by enkelmedia** [comment]
Data Type configuration variants :) I've never missed this, but after reading the proposal I realized that we might have a project in the near future where we'd need a different color palette depending on the root note.

Not sure that this is a super common use case but I would really like to know more about how @nackler solve it?

**Reply by nackler** [reply]
If you have multiple root nodes, then it seems like you could add a color property to each one then on any view, call
something like this:
```

 <div style="background-color:@ViewUtility.GetColorForRootNode(this.Umbraco)"><div>

public static class ViewUtility{
    public static GetColorForRootNode(UmbracoHelper umbracoHelper){
        var rootNodeForCurrentPage = (MyRootNodePageType)umbracoHelper.AssignedContentItem.Root();
        return rootNodeForCurrentPage.MyColorProperty
    }
}
```

**Reply by enkelmedia** [reply]
Thanks! But this would not solve the problem with allowing editors to pick different colors eg. If I have a color picker on a block and would like the pickable colors to be different depending on the root node.

**Reply by hfloyd** [reply]
@enkelmedia - Oops - put it into the wrong 'thread"... but take a look here - https://github.com/umbraco/Umbraco-CMS/discussions/17216#discussioncomment-11350542

**Reply by clementfavre** [reply]
@hfloyd that would be perfect! that's what I wanted to do in the first place, could you show me in more detail how you came up with it?

**Reply by hfloyd** [reply]
@D0LBA3B - Sure. Here is an example which works in Umbraco 10-13. (Some changes might be needed for 14+)

So, the first step is to determine where you are storing the lists of colors, and how you link them to the "sites". In my case, I'm using my theming package, so each site is assigned a "theme name" (string). There is a matching named folder which includes a JSON file of the colors. In your case, you'd need to decide where your list will be stored - whether in a file, or in a content node or whatever.


Next, you'll need to code a [Contentment "custom data source"](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/data-list.md#extending-with-your-own-custom-data-source) where you use the current node to retrieve the correct colors list, and transform it into a list of `DataListItem`s. This is a simple object with 4 string properties.

In my case, using the JSON file, I have the colors already defined using this format, so in my custom data source, I just get the theme name from the site root node, grab the file from the correct folder, and deserialize the JSON directly into this model

Example:

```json
...
 {
    "name": "Gray",
    "value": "#A9B3BA",
    "icon": "icon-palette",
    "description": "gray"
 },
 {
   "name": "White",
   "value": "#FFFFFF",
   "icon": "icon-palette",
   "description": "white"
 },
 {
   "name": "Black",
   "value": "#000000",
   "icon": "icon-palette",
   "description": "black"
 }
...
```
The `value` property will be what is stored when you select the color. In this case, it is storing a HEX code, but you could use a class name or some other value, as desired. One of the properties will need to be a HEX code, though, if you want a visible color swatch in the UI.

Then you can create a new Datatype using Contentment DataList with your new custom source as the "Data source", and for the "List editor" choose "Templated List". This is my template code (based on the property structure in the example above):

```html
<style>
.color-box{
width:100%;
height:100%;
}
</style>
 <div class="umb-action" 
            ng-class="{'selected': item.selected == true}" 
            style="border: 1px solid #a2a1a6; margin: 2px; cursor:pointer; display: inline-block; padding:4px 6px; width:80px; height:120px;">
    <div style="height:30px;">
        <div class="color-box" ng-style="{'background-color': item.value}"></div>
    </div>
    <div style="height:60px;margin:4px 2px;" 
        ng-bind="item.name"></div>
    <div style="height:20px;margin:0 2px;font-size:60%;vertical-align:bottom;line-height: normal;" >
            <span ng-bind="item.description"></span> | <span  ng-bind="item.value"></span>
    </div>
</div>
```
Based on your structure, you might need to use different properties for the color-box, etc. 

Also, add a single "HTML attribute" like this:
Name: `class`
Value: `umb-sortable-thumbnails-container umb-property-editor`

This will result in a UI like this:
![image](https://github.com/user-attachments/assets/1ef7f9a9-6e69-4efe-8c2c-67be708b1546)

Now you can use this DataType on any properties which need a site-specific color picker.

**Reply by bielu** [reply]
@hfloyd Woudn't be better option to use relative Settings node to root node to retrieve colours?(assuming you can store info in umbraco instead of file). you can quite easily achieve it with contement from repeatable string or nested content with colour eye picker/colour picker (Idatalistsource / IDataPickerSource) 🤔 I done multiple crazy things with contement and it is pretty flexible about source of data! so depends on requirements json might not be best option 🤔

**Reply by hfloyd** [reply]
The "Best" option is the one which meets the various project requirements. As I mentioned at the beginning, there are plenty of options for WHERE the data can be stored - and it depends on what makes the most sense for the project. The point is that Contentment DataList can be used to retrieve the list of colors, from wherever it is residing, and that list should be convertible into the expected 4-property structure. Considering that JSON is easy to edit, and generally the color list is determined at the time of the design - not willy-nilly after the fact, I found that it was a convenient option. YMMV

**Comment by marcemarc** [comment]
@enkelmedia with V13 + Contentment you could create a Pretty Contentment DataList with a custom IDataListSource, and the trick was to use the EditorModel SendingContent notification to place the current page ID into a session variable (I know!) - then the custom IDataListSource could read the value of that ID from the session and use it to find the site the page was being created on, and read a set of brand colours on that homepage as the ones to use as the custom data list source for editors to select from... The same block can be used across multiple sites and the colour options for the editor are updated to reflect the context of the site the page is created on. 

But have been trying to think how to do something similar in v14+ without the EditorModel SendingContent hack..

**Reply by enkelmedia** [reply]
I love it! Thanks for sharing your awesome hack!

This is interesting, I’ll try to think of a way - and for sure share it if I come up with something. Not actively working on it at the moment so might take a while.

**Reply by leekelleher** [reply]
It could be done with [Contentment](https://github.com/leekelleher/umbraco-contentment) without the hack. 😗👌 

Using the [**List Items** editor](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/list-items.md) on a site/root document to define the colour values. Then configure a [**Data List** editor](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/data-list.md) to use the [**Umbraco Content Property Values** data-source](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/data-sources/data-source--umbraco-content-property-values.md), configured to query the List Item values from the site/root document.

...and if you don't want to use one of the built-in Data List list-editors, then custom markup (AngularJS template) can be used with the **Templated List** list-editor, or [roll your own custom list-editor](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/data-list.md#extending-with-your-own-custom-list-editor).

**Reply by marcemarc** [reply]
> I love it! Thanks for sharing your awesome hack!
> 
> This is interesting, I’ll try to think of a way - and for sure share it if I come up with something. Not actively working on it at the moment so might take a while.

A @hfloyd hack!

**Reply by marcemarc** [reply]
> It could be done with [Contentment](https://github.com/leekelleher/umbraco-contentment) without the hack. 😗👌
> 
> Using the [**List Items** editor](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/list-items.md) on a site/root document to define the colour values. Then configure a [**Data List** editor](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/data-list.md) to use the [**Umbraco Content Property Values** data-source](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/data-sources/data-source--umbraco-content-property-values.md), configured to query the List Item values from the site/root document.
> 
> ...and if you don't want to use one of the built-in Data List list-editors, then custom markup (AngularJS template) can be used with the **Templated List** list-editor, or [roll your own custom list-editor](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/data-list.md#extending-with-your-own-custom-list-editor).

ooh fixed indeed in V5 of contentment - https://github.com/leekelleher/umbraco-contentment/pull/389 - so that should work for Umbraco V15?

**Reply by leekelleher** [reply]
Contentment v5 is for Umbraco v13.
Contentment v6 will be for Umbraco v15 _(which I'm working on at this very moment), the previous v6 alphas did target Umbraco v14._

re: `ContentmentContentContext`, that might not be needed in v6, as I've re-engineered Data List to work a little differently, (again still working on that at the moment). 😅

**Reply by marcemarc** [reply]
Thanks @leekelleher it's one of my 'todo slides' for 'who grated my cheese?'

![image](https://github.com/user-attachments/assets/6d3b6f88-3cc8-414b-bf26-87b4e4755798)

so have been pondering 'how'

**Reply by hfloyd** [reply]
I have a way of doing just this - having multiple sites with their own defined lists of colors for selecting. I feel like I've written it up before, but maybe not... In my version, I'm using JSON files to define the lists of colors, and the site's node provides the "theme" name, which determines which JSON file is used to supply the color list to a Contentment Datalist, with a custom template providing a UI like this:
![image](https://github.com/user-attachments/assets/2b80f6a1-3cc0-4fdb-87fb-96ce3b1b8faf)
(Multiple sites - same installation)

![image](https://github.com/user-attachments/assets/fa6e8512-e0c7-47d4-a850-f1f2e9a1f4b0)
![image](https://github.com/user-attachments/assets/c8b2a4e4-9b75-4a87-a40d-a30428d577be)

If this is what you are looking for, I can elaborate.

**Reply by hfloyd** [reply]
Ps. HERE is the detailed explanation: https://github.com/umbraco/Umbraco-CMS/discussions/17216#discussioncomment-11450764
